stdlib/public/runtime/HeapObject.cpp

```diff
@@ -78,6 +79,34 @@ HeapObject *swift::swift_allocObject(HeapMetadata const *metadata,
   return _swift_allocObject(metadata, requiredSize, requiredAlignmentMask);
 }
 
+#if OBJC_SETASSOCIATEDOBJECTHOOK_DEFINED
+//We interpose objc_setAssociatedObject so that we can set a flag in
+//the refcount field of Swift objects to indicate that they have associations,
+//since we can't safely skip ObjC dealloc work if they do
+static objc_hook_setAssociatedObject originalAssocObjectFunc = nullptr;
+
+static void _swift_setAssociatedObject_hook(
+  id _Nonnull object,
+  const void * _Nonnull key,
+  id _Nullable value,
+  objc_AssociationPolicy policy
+) {
+  if (!isObjCTaggedPointerOrNull(object) &&
+      objectUsesNativeSwiftReferenceCounting(object)) {
+    auto heapObj = reinterpret_cast<HeapObject *>(object);
+    heapObj->refCounts.setPureSwiftDeallocation(false);
+  }
+  originalAssocObjectFunc(object, key, value, policy);
+}
+
+static void _interpose_objc_association(void *ctxt) {
+  if (__builtin_available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *))) {
```

Comment:
This is new, and I've verified that it compiles to nothing when targeting the listed OSs, so the in-OS build shouldn't see any overhead from it.